hw1/task1/main.py

```diff
@@ -1,99 +1,74 @@
-
-
-def missing_dates(path=None, data=None):
-    if data is None and path is not None:
-        data = pd.read_csv(path)
-    elif path is None and data is not None:
-        data = data
-    else:
-        return None
-
-    missing_data_indices = np.append(np.where(data['Passenger_count'] == 0),
-                                     np.unique(np.append(np.where(data['Pickup_longitude'] == 0),
-                                               np.where(data['Dropoff_longitude'] == 0))))
-    missing_dates = pd.DataFrame(data['Lpep_dropoff_datetime'][missing_data_indices])
-    missing_dates.columns = ['missing_dates']
-    missing_dates.to_csv('missing_dates.csv', index=False)
-
-    return missing_dates, data
-
-
-def usage_stat(path=None, data=None):
-    if data is None and path is not None:
-        data = pd.read_csv(path)
-    elif path is None and data is not None:
-        data = data
-    else:
-        return None
-    invalid_indices = np.logical_or(data['Pickup_longitude'] == 0, data['Dropoff_longitude'] == 0)
-    data['Lpep_dropoff_datetime'] = pd.to_datetime(data['Lpep_dropoff_datetime'])
-    valid_indices = np.logical_not(invalid_indices)
-    stat = data[np.logical_not(invalid_indices)].groupby(data['Lpep_dropoff_datetime'].dt.date)['VendorID'].count()
-    stat.index.name = 'date'
-    stat = pd.DataFrame(stat)
-    stat.columns = ['count']
-    stat.to_csv('usage_stat.csv')
-    return stat, data
-
-
-def trip_stat(path=None, data=None):
-    if data is None and path is not None:
-        data = pd.read_csv(path)
-    elif path is None and data is not None:
-        data = data
+import glob
+import subprocess
+import get_stats as stat
+import averaging as avrg
+import pandas as pd
+import multiprocessing
+
+
+def calc(path):
+    try:
+        general_stat = stat.general_stats(path)
+    except ValueError as v:
+        if subprocess.call(['sed', '-i', 's/,,,//', path]) == 0:
+            general_stat, data = stat.general_stats(path)
+        else:
+            raise v
+    missing_dates = stat.missing_dates(path)
+    usage_stat = stat.usage_stat(path)
+    trip_stat = stat.trip_stat(path)
+
+    return general_stat, missing_dates, usage_stat, trip_stat
+
+
+def many_files(paths):
+    if len(paths) == 0:
+        raise FileExistsError('There is no files')
+
+    general_stats = pd.DataFrame(columns=['mean_cost',
+                                          'longest_ride',
+                                          'max_count',
+                                          'max_count_start',
+                                          'max_count_end',
+                                          'invalid_rows'])
+    missing_stats = pd.DataFrame(columns=['missing_dates'])
+    usage_stats = pd.DataFrame(columns=['count'])
+    trip_stats = pd.DataFrame(columns=['trip_distance'])
+    results = []
+
+    p = multiprocessing.Pool(processes=4)
+    for path in paths:
+        async_res = p.map_async(calc, [path])
```

Comment:
why do you start each path separately?
as far as I understand, you can specify the list in map_async directly